### PR TITLE
Feature/base generator

### PR DESCRIPTION
**Pull request by faneshion**
#217  An initial version of base generator. It need more discussion.

reference to #245 .

### CODE REVIEW

File: matchzoo/engine/base_generator.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
+"""Base generator."""
+
+import abc
+import keras
+import threading
+import numpy as np
+
+
+class BaseGenerator(keras.utils.Sequence):
```
Comment (pl8787) [inline matchzoo/engine/base_generator.py]: Why ```BaseGenerator``` inherits from ```keras.utils.Sequence```?

File: matchzoo/generators/point_generator.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        num_class: int=2,
+        text_right_max_len: int=5,
+        text_left_max_len: int=100,
+        batch_size: int=32,
+        shuffle: bool=True,
+        is_train: bool=False
+    ):
+        """Initialization of point-generator."""
+        self._text_left_max_len = text_left_max_len
+        self._text_right_max_len = text_right_max_len
```
Comment (pl8787) [inline matchzoo/generators/point_generator.py]: Leaving the text trim operation in preprocessing step?

File: matchzoo/generators/point_generator.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+            x1 = sample[1]
+            y = sample[2]
+            x0_id = sample[3]
+            x1_id = sample[4]
+            ctl = min(self.text_left_max_len, len(x0))
+            ctr = min(self.text_right_max_len, len(x1))
+            xx0 = np.zeros((self.text_left_max_len,), dtype=np.int32)
+            xx1 = np.zeros((self.text_right_max_len,), dtype=np.int32)
+            xx0[:] = fill_word   # padding
+            xx1[:] = fill_word   # padding
```
Comment (pl8787) [inline matchzoo/generators/point_generator.py]: Leaving the padding operation in preprocessing step?

File: matchzoo/engine/base_generator.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        """Gets a batch of transformed samples.
+
+        # Arguments
+            index_array: Arrray of sample indices to include in batch.
+
+        # Returns
+            A batch of transformed samples.
+
+        """
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (faneshion) [reply matchzoo/engine/base_generator.py]: ok

File: matchzoo/generators/point_generator.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        batch_ids = []
+        if self.task_type == tasks.Regression:
+            batch_y = self.y
+        elif self.target_mode == tasks.Classification:
+            batch_y = np.zeros((batch_size, self.task_type._num_classes),
+                               dtype=np.int32)
+            for i, label in enumerate(self.y[index_array]):
+                batch_y[i, label] = 1
+        else:
+            raise ValueError('Error target mode in point generator.')
```
Comment (bwanglzu) [inline matchzoo/generators/point_generator.py]: raise ValueError("{} is not a valid task type".format(task_type))

File: matchzoo/generators/point_generator.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+        for i, j in enumerate(index_array):
```
Comment (bwanglzu) [inline matchzoo/generators/point_generator.py]: `i` `j` , `index_array` is extremely unreadble 
`for key, value in enumerate(indicies_of_current_batch):`

File: matchzoo/generators/point_generator.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            batch_x_left[i] = self.x_left[j]
+            batch_x_right[i] = self.x_right[j]
```
Comment (bwanglzu) [inline matchzoo/generators/point_generator.py]: same, replace `i`, `j`

File: tests/unit_test/generators/test_point_generator.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+@pytest.fixture(scope='module', params=[
+    tasks.Classification(num_classes=2),
+    tasks.Classification(num_classes=16),
+    tasks.Ranking()
+])
+def task(request):
+    return request.param
+
+@pytest.fixture
+def raw_point_generator(x, task):
```
Comment (uduse) [inline tests/unit_test/generators/test_point_generator.py]: Is this a fixture or a test case?

File: matchzoo/generators/point_generator.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    Ponit generator can be used for classification as well as ranking.
+
+    # Arguments
+        inputs: the transformed dataset
+    """
+
+    def __init__(
+        self,
+        inputs: DataPack,
+        task_type: engine.BaseTask,
```
Comment (uduse) [inline matchzoo/generators/point_generator.py]: `task` is more appropriate than `task_type` since this parameter is passed as an instance of a task type.

File: matchzoo/generators/point_generator.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        """Initialize the point generator."""
+        self.task_type = task_type
+        self.n = len(inputs)
+        self.transform_data(inputs)
+        super(PointGenerator, self).__init__(batch_size, shuffle)
+
+    def transform_data(self, inputs: DataPack):
+        """Obtain the transformed data from datapack."""
+        data = inputs.dataframe
+        self.x_left = np.asarray(data.text_left)
```
Comment (uduse) [inline matchzoo/generators/point_generator.py]: Shouldn't add any attributes outside of `__init__`.

File: matchzoo/engine/base_generator.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    def __init__(self, batch_size: int, shuffle: bool):
+        """Initialize the base generator."""
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.batch_index = 0
+        self.index_array = None
+        self.index_generator = self._flow_index()
+
+    def _set_index_array(self):
+        self.index_array = np.arange(self.n)
```
Comment (bwanglzu) [inline matchzoo/engine/base_generator.py]: where did you created `self.n`?

File: matchzoo/engine/base_generator.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        if self.shuffle:
+            self.index_array = np.random.permutation(self.n)
+
+    def __getitem__(self, idx):
+        """Get a batch from index `idx`."""
+        if idx >= len(self):
+            raise ValueError('Asked to retrieve element {idx}, '
+                             'but the Sequence '
+                             'has length {length}'.format(idx=idx,
+                                                          length=len(self)))
```
Comment (bwanglzu) [inline matchzoo/engine/base_generator.py]:
```python
msg  = f'Asked to retrieve element at index {idx}'
msg += f'but the Sequence has length {len(self)}'
raise ValueError(msg)
```
Reference python  f-strings: https://realpython.com/python-f-strings/

File: matchzoo/engine/base_generator.py [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    def _get_batch_of_transformed_samples(self, index_array):
+        """Get a batch of transformed samples.
+
+        # Arguments
+            index_array: Arrray of sample indices to include in a batch.
+
+        # Returns
+            A batch of transformed samples.
+
+        """
```
Comment (bwanglzu) [inline matchzoo/engine/base_generator.py]: docstring:
```python
def _get_batch_of_transformed_samples(self, index_array: list):
    """
    Get a batch of transformed samples.
    
   :param index_array: List of sample indices to .....
   :return: A batch of ....
    """
```

File: matchzoo/engine/base_generator.py [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+import abc
+import keras
+import numpy as np
+
+
+class BaseGenerator(keras.utils.Sequence):
+    """Abstract base generator of all matchzoo generators.
+
+    Every `Generator` must implement the `_get_batch_of_transformed_samples`
```
Comment (bwanglzu) [inline matchzoo/engine/base_generator.py]: Every `Generator` must implement the :meth:\`_get_batch_of_transformed_samples\`

`:meth:` represents for a method, just like `:class:` and `:attr:` is used to generate docstrings.

File: matchzoo/engine/base_generator.py [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    method.
+
+    # Arguments
+        batch_size: Integer, size of a batch.
+        shuffle: Boolean, whether to shuffle the data between epochs.
```
Comment (bwanglzu) [inline matchzoo/engine/base_generator.py]: remove line 14-16.

File: matchzoo/engine/base_generator.py [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        index_array = self.index_array[self.batch_size * idx:
+                                       self.batch_size * (idx + 1)]
+        return self._get_batch_of_transformed_samples(index_array)
+
+    def __len__(self):
+        """Return the total number of batches."""
+        return (self.n + self.batch_size - 1) // self.batch_size
+
+    def on_epoch_end(self):
+        """Reorganize the index array while epoch is ended."""
```
[the remark this answers is not in the document]
Comment (faneshion) [reply matchzoo/engine/base_generator.py]: reorganize = re-organize~

File: matchzoo/generators/point_generator.py (hunk 10 above)
Comment (faneshion) [reply matchzoo/generators/point_generator.py]: So, can I assign the returned value to these attributes? just like `self.text_left, self.text_right = self.transform_data(inputs)`.

File: matchzoo/generators/point_generator.py (hunk 9 above)
Comment (faneshion) [reply matchzoo/generators/point_generator.py]: ok~

File: tests/unit_test/generators/test_point_generator.py (hunk 8 above)
Comment (faneshion) [reply tests/unit_test/generators/test_point_generator.py]: It's a fixture~

File: matchzoo/engine/base_generator.py (hunk 11 above)
Comment (faneshion) [reply matchzoo/engine/base_generator.py]: It is created in `PointGenerator`.

File: matchzoo/engine/base_generator.py (hunk 12 above)
Comment (faneshion) [reply matchzoo/engine/base_generator.py]: Nice~

File: matchzoo/engine/base_generator.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (bwanglzu) [reply matchzoo/engine/base_generator.py]: ...

File: tests/unit_test/generators/test_point_generator.py (hunk 8 above)
Comment (uduse) [reply tests/unit_test/generators/test_point_generator.py]: Then how come no one's using it?

File: matchzoo/generators/point_generator.py (hunk 5 above)
Comment (bwanglzu) [reply matchzoo/generators/point_generator.py]: @faneshion I guess the comment is more reable `raise ValueError("{} is not a valid task type".format(task_type))`

File: tests/unit_test/generators/test_point_generator.py (hunk 8 above)
Comment (bwanglzu) [reply tests/unit_test/generators/test_point_generator.py]: same question
